librz/il/vm_layer.c

```diff
@@ -0,0 +1,339 @@
+	bind_options.calcsizeK = (HtPPCalcSizeK)strlen;
+	vm->vm_global_bind_table = ht_pp_new_opt(&bind_options);
+
+	// TODO : More Arguments for vm init to control
+	//      1. Minimal unit size in memory
+	//      2. Multiple Memory
+	//      3. pc length
+	vm->mems = (RzILMem **)calloc(RZ_IL_VM_MAX_TEMP, sizeof(RzILMem *));
+	if (!vm->mems) {
+		RZ_LOG_ERROR("[VM INIT FAILED] : mem\n");
+		rz_il_vm_close(vm);
+		return false;
+	}
+	vm->pc = rz_il_bv_new_from_ut64(addr_size, start_addr);
+
+	// Table for storing the core theory opcodes
+	HtPPOptions ops_options = { 0 };
+	ops_options.cmp = (HtPPListComparator)rz_il_bv_cmp;
+	ops_options.hashfn = (HtPPHashFunction)rz_il_bv_hash;
+	ops_options.dupkey = (HtPPDupKey)rz_il_bv_dup;
+	ops_options.dupvalue = NULL; // dump key only, since the opcode used in hash map only
+	ops_options.freefn = free_opcode_kv;
+	ops_options.elem_size = sizeof(HtPPKv);
+	vm->ct_opcodes = ht_pp_new_opt(&ops_options);
+
+	// init jump table of labels
+	vm->op_handler_table = RZ_NEWS0(RzILOpHandler, RZIL_OP_MAX);
+	vm->op_handler_table[RZIL_OP_VAR] = &rz_il_handler_var;
+	vm->op_handler_table[RZIL_OP_ITE] = &rz_il_handler_ite;
+	vm->op_handler_table[RZIL_OP_UNK] = &rz_il_handler_unk;
+
+	vm->op_handler_table[RZIL_OP_B0] = &rz_il_handler_b0;
+	vm->op_handler_table[RZIL_OP_B1] = &rz_il_handler_b1;
+	vm->op_handler_table[RZIL_OP_AND_] = &rz_il_handler_and_;
+	vm->op_handler_table[RZIL_OP_OR_] = &rz_il_handler_or_;
+	vm->op_handler_table[RZIL_OP_INV] = &rz_il_handler_inv;
+
+	vm->op_handler_table[RZIL_OP_LOAD] = &rz_il_handler_load;
+	vm->op_handler_table[RZIL_OP_STORE] = &rz_il_handler_store;
+
+	vm->op_handler_table[RZIL_OP_INT] = &rz_il_handler_int;
+	vm->op_handler_table[RZIL_OP_NEG] = &rz_il_handler_neg;
+	vm->op_handler_table[RZIL_OP_NOT] = &rz_il_handler_not;
+	vm->op_handler_table[RZIL_OP_LSB] = &rz_il_handler_lsb;
+	vm->op_handler_table[RZIL_OP_MSB] = &rz_il_handler_msb;
+	vm->op_handler_table[RZIL_OP_SHIFTL] = &rz_il_handler_shiftl;
+	vm->op_handler_table[RZIL_OP_SHIFTR] = &rz_il_handler_shiftr;
+	vm->op_handler_table[RZIL_OP_ADD] = &rz_il_handler_add;
+	vm->op_handler_table[RZIL_OP_SUB] = &rz_il_handler_sub;
+	vm->op_handler_table[RZIL_OP_MUL] = &rz_il_handler_mul;
+	vm->op_handler_table[RZIL_OP_DIV] = &rz_il_handler_div;
+	vm->op_handler_table[RZIL_OP_MOD] = &rz_il_handler_mod;
+	vm->op_handler_table[RZIL_OP_SDIV] = &rz_il_handler_sdiv;
+	vm->op_handler_table[RZIL_OP_SMOD] = &rz_il_handler_smod;
+
+	vm->op_handler_table[RZIL_OP_PERFORM] = &rz_il_handler_perform;
+	vm->op_handler_table[RZIL_OP_SET] = &rz_il_handler_set;
+	vm->op_handler_table[RZIL_OP_GOTO] = &rz_il_handler_goto;
+	vm->op_handler_table[RZIL_OP_BRANCH] = &rz_il_handler_branch;
+	vm->op_handler_table[RZIL_OP_SEQ] = &rz_il_handler_seq;
+
+	vm->var_count = 0;
+	vm->val_count = 0;
+	vm->mem_count = 0;
+	vm->easy_debug = 0;
+
+	return true;
+}
+
+/**
+ * Close and clean vm
+ * \param vm RzILVM* pointer to VM
+ */
+RZ_API void rz_il_vm_close(RzILVM *vm) {
```

Comment:
We don't usually use `close`. Also, there is no `open` AFAIK, so I think it's better to just use `init`/`fini` as usual. And given that the `init` does not allocate the memory for RzILVM, the `fini` should not free the `vm` itself.